---
title: "Arctic picos"
output:
  html_document:
    df_print: paged
---

This is an [R Markdown](http://rmarkdown.rstudio.com) Notebook. When you execute code within the notebook, the results appear beneath the code. 

Try executing this chunk by clicking the *Run* button within the chunk or by placing your cursor inside it and pressing *Ctrl+Shift+Enter*. 

```{r packages, echo=TRUE, warning=FALSE}
require(iNEXT); packageVersion("iNEXT")
require(tidyverse); packageVersion("tidyverse")
require(tidyr); packageVersion("tidyr")
require(dplyr); packageVersion("dplyr")
require(ggplot2); packageVersion("ggplot2")
require(vegan); packageVersion("vegan")
require(tibble) #for "rownames_to_column('Site')" 
require(reshape2); packageVersion("reshape2")
library(dada2)
packageVersion("dada2")
library(ShortRead) #(for intallation with Biocmanager use the lib="~/R/x86_64-pc-linux-gnu-library/4.3" command)
#packageVersion("ShortRead")
library(Biostrings)
#packageVersion("Biostrings")
library(stringr)
```

In qiime2

#convert dada2 output to qiime biom object
qiime tools import \
--input-path ./ArcticPicos16S_rep-seqs.fna \
--type 'FeatureData[Sequence]' \
--output-path ./ArcticPicos_16S_rep-seqs.qza

echo -n "#OTU Table" | cat - ./ArcticPicos_16S_seqtab-nochim.txt > ./dada2-analysis/biom-table.txt

biom convert -i dada2-analysis/biom-table.txt -o dada2-analysis/table.biom --table-type="OTU table" --to-hdf5

cd  dada2-analysis

qiime tools import \
  --input-path ./dada2-analysis/table.biom \
  --type 'FeatureTable[Frequency]' \
  --output-path ./dada2-analysis/table.qza

qiime picrust2 full-pipeline \
   --i-table ./dada2-analysis/table.qza \
   --i-seq ArcticPicos_16S_rep-seqs.qza \
   --output-dir q2-picrust2_output_mp \
   --p-placement-tool epa-ng \
   --p-threads 1 \
   --p-hsp-method mp \
   --p-max-nsti 2 \
   --verbose
   
qiime feature-table summarize \
   --i-table ./q2-picrust2_output/pathway_abundance.qza \
   --o-visualization ./q2-picrust2_output/pathway_abundance.qzv


#check output results with
 qiime tools view ./q2-picrust2_output/pathway_abundance.qzv
 
#min 735585

qiime diversity core-metrics \
   --i-table ./q2-picrust2_output/pathway_abundance.qza \
   --p-sampling-depth 735585 \
   --m-metadata-file meta_16S_functional.tsv \
   --output-dir ./pathabun_core_metrics_out \
   --p-n-jobs 1

qiime tools export --input-path q2-picrust2_output/ec_metagenome.qza --output-path metaG_exported
   
biom convert \
   -i pathabun_exported/feature-table.biom \
   -o pathabun_exported/feature-table.biom.tsv \
   --to-tsv


```{r import_data, echo=TRUE, warning=FALSE}
source("/Users/corahoerstmann/Documents/AWI_ArcticFjords/Code/data_import/import.R", encoding = "UTF-8")

#source("/Users/corahoerstmann/Documents/AWI_ArcticFjords/Code/data_import/rename_metadata.R", #encoding = "UTF-8")

```

16S: DOI
10.5281/zenodo.4587955

### remove replicates and outliers

outliers arose from crysochromulina bloom (John et al. in prep) and replicates only available from HE533, thus we always used Rep. A from each site.


```{r remove_replicates, echo=TRUE, warning=FALSE}
outliers_euks <- c("HE533.Euk.F02.04C_S9", "HE533.Euk.F02.08B_S20", "HE533.Euk.F02.08C_S21", "HE533.Euk.F02.09A_S22",
              "HE533.Euk.F02.10A_S25", "HE533.Euk.F02.10C_S27", "HE533.Euk.F02.11A_S28", "HE533.Euk.F02.11C_S30",
              "HE533.Euk.F02.12B_S32", "HE533.Euk.F02.12C_S33", "HE533.Euk.F02.19C_S54", "HE533.Euk.F02.22C_S63")

#remove replicates

rep_18S <- c("HE533.Euk.F02.25B_S68", "HE533.Euk.F02.25C_S69", "HE533.Euk.F02.26B_S71", "HE533.Euk.F02.26C_S72", "HE533.Euk.F02.27B_S74", "HE533.Euk.F02.27C_S75", "HE533.Euk.F02.28B_S77", "HE533.Euk.F02.28C_S78",
"HE533.Euk.F02.21B_S59", "HE533.Euk.F02.21C_S60", "HE533.Euk.F02.22B_S62", "HE533.Euk.F02.23B_S65", "HE533.Euk.F02.23C_S66", "HE533.Euk.F02.17B_S47", "HE533.Euk.F02.17C_S48", "HE533.Euk.F02.18B_S50", "HE533.Euk.F02.18C_S51", "HE533.Euk.F02.19B_S53", "HE533.Euk.F02.20B_S56", "HE533.Euk.F02.20C_S57",
"HE533.Euk.F02.13C_S36", "HE533.Euk.F02.14B_S38", "HE533.Euk.F02.14C_S39", "HE533.Euk.F02.15B_S41", "HE533.Euk.F02.15C_S42", "HE533.Euk.F02.16B_S44", "HE533.Euk.F02.16C_S45", "HE533.Euk.F02.07B_S17", "HE533.Euk.F02.07C_S18", "HE533.Euk.F02.09C_S24", "HE533.Euk.F02.04B_S8", "HE533.Euk.F02.05B_S11",
"HE533.Euk.F02.05C_S12", "HE533.Euk.F02.06B_S14", "HE533.Euk.F02.06C_S15", "HE533.Euk.F02.2B_S2", "HE533.Euk.F02.2C_S3", "HE533.Euk.F02.3B_S5", "HE533.Euk.F02.3C_S6")

rep_16S <- c("HE533.Prok.F02.25B_S90",  "HE533.Prok.F02.25C_S91", "HE533.Prok.F02.26B_S94", "HE533.Prok.F02.26C_S95", "HE533.Prok.F02.27B_S98", "HE533.Prok.F02.27C_S99", "HE533.Prok.F02.28B_S102", "HE533.Prok.F02.28C_S103", "HE533.Prok.F02.21B_S78",  "HE533.Prok.F02.21C_S79", "HE533.Prok.F02.22B_S82",  "HE533.Prok.F02.22C_S83", "HE533.Prok.F02.23B_S86", "HE533.Prok.F02.23C_S87", "HE533.Prok.F02.17B_S62", "HE533.Prok.F02.17C_S63", "HE533.Prok.F02.18B_S66",  "HE533.Prok.F02.18C_S67", "HE533.Prok.F02.19B_S70",  "HE533.Prok.F02.19C_S71", "HE533.Prok.F02.20B_S74", "HE533.Prok.F02.20C_S75", "HE533.Prok.F02.11B_S38",  "HE533.Prok.F02.11C_S39", "HE533.Prok.F02.12B_S42",  "HE533.Prok.F02.12C_S43", "HE533.Prok.F02.13B_S46", "HE533.Prok.F02.13C_S47", "HE533.Prok.F02.14B_S50", "HE533.Prok.F02.14C_S51", "HE533.Prok.F02.15B_S54",  "HE533.Prok.F02.15C_S55", "HE533.Prok.F02.16B_S58",  "HE533.Prok.F02.16C_S59", "HE533.Prok.F02.10B_S22",  "HE533.Prok.F02.7B_S34",  "HE533.Prok.F02.7C_S23", "HE533.Prok.F02.8B_S26",   "HE533.Prok.F02.8C_S27", "HE533.Prok.F02.2B_S2", "HE533.Prok.F02.2C_S3", "HE533.Prok.F02.3B_S6", "HE533.Prok.F02.3C_S7", "HE533.Prok.F02.4B_S10", "HE533.Prok.F02.4C_S11", "HE533.Prok.F02.5B_S14", "HE533.Prok.F02.5C_S15", "HE533.Prok.F02.6B_S18", "HE533.Prok.F02.6C_S19", "HE533.Prok.F02.9B_S30", "HE533.Prok.F02.9C_S31")


###remove outliers

euk_r <- eukaryotes%>%dplyr::select(!outliers_euks)
euk_r <- euk_r%>%dplyr::select(!rep_18S)
euk_r <- euk_r[rowSums(euk_r)>0,]

prok_r <- prokaryotes%>%dplyr::select(!rep_16S)
prok_r <- prok_r[rowSums(prok_r)>0,]
prokaryotes = prok_r

meta_18S_r <- meta_18S%>%dplyr::filter(Site %in% colnames(euk_r))
meta_16S <- meta_16S%>%dplyr::filter(Site %in% colnames(prokaryotes))

taxonomy_16S_r <- taxonomy_16S%>%filter(rownames %in% rownames(prok_r))
rm(outliers_euks, rep_16S, rep_18S)
```

prok_r$rownames <- rownames(prok_r)
prokaryotes_F <- left_join(prok_r, taxonomy_16S[,c(9,7)], by = "rownames")
rownames(prokaryotes_F) <- prokaryotes_F$sequence
prokaryotes_F$sequence <- NULL
prokaryotes_F$rownames <- NULL
prokaryotes_F <- t(prokaryotes_F)
#to use this in qiime (https://forum.qiime2.org/t/exporting-dada2-r-package-results-to-work-with-qiime2/2573/5)
uniquesToFasta(prokaryotes_F, fout='/Users/corahoerstmann/Documents/AWI_ArcticFjords/Datasets/16S_functional/ArcticPicos16S_rep-seqs.fna', ids=colnames(prokaryotes_F))
write.table(t(prokaryotes_F), "/Users/corahoerstmann/Documents/AWI_ArcticFjords/Datasets/ArcticPicos_16S_seqtab-nochim.txt", sep="\t", row.names=TRUE, col.names=NA, quote=FALSE)
####

write_tsv(meta_16S, "meta_16S_functional.tsv")

import the distance tables of drifters

```{r drifter_model, echo=TRUE, warning=FALSE}
source("/Users/corahoerstmann/Documents/AWI_ArcticFjords/Code/env_calc/drifter_calc.R", encoding = "UTF-8")

list_drifter <- sapply(ls(pattern="drifter_"), function(x) get(x), simplify = FALSE)
rm(list = ls(pattern="drifter_"))
rm(info)
```


### metadata removing NAs and z-scoring 
```{r z_score, echo=FALSE, warning = FALSE}
source("/Users/corahoerstmann/Documents/AWI_ArcticFjords/Code/env_calc/meta_reduction.R", encoding = "UTF-8")
```

## metadata exploration

```{r meta, echo=TRUE}
source("/Users/corahoerstmann/Documents/AWI_ArcticFjords/Code/env_calc/MLD_DK.R", encoding = "UTF-8")
source("/Users/corahoerstmann/Documents/AWI_ArcticFjords/Code/env_calc/MLD_calculations.R", encoding = "UTF-8")

```


add seasonality (based on sun altitude and azimuth) to the analysis 

```{r suncalc, echo=FALSE}
source("/Users/corahoerstmann/Documents/AWI_ArcticFjords/Code/env_calc/SunAngleCalc.R", encoding = "UTF-8")
meta_sun2_16S <- Sunanglecalc(meta_16S_m)
meta_sun2_18S <- Sunanglecalc(meta_18S_m)

meta_sun2_16S%>%group_by(Fjord2)%>%summarise(min_alt = min(altitude), max_alt = max(altitude), min_az = min(azimuth), max_az = max(azimuth))
meta_sun2_18S%>%group_by(Fjord2)%>%summarise(min_alt = min(altitude), max_alt = max(altitude), min_az = min(azimuth), max_az = max(azimuth))

meta_16S_m <- dplyr::inner_join(meta_16S_m, meta_sun2_16S[,c("Site","altitude", "azimuth")], by = c("Site"))

meta_18S_m <- dplyr::inner_join(meta_18S_m, meta_sun2_18S[,c("Site","altitude", "azimuth")], by = c("Site"))

rm(meta_sun2_16S, meta_sun2_18S)

```

create list of metadata
```{r}

meta_all <- dplyr::inner_join(meta_16S_m, meta_18S_nNA[1:2], by="Station")
meta_all$In.Out <- as.factor(meta_all$In.Out)

meta_all%>%group_by(Fjord2)%>%summarise(min_depth = min(bottom_depth), max_depth = max(bottom_depth))

list_meta <- sapply(ls(pattern="meta_"), function(x) get(x), simplify = FALSE)
rm(list = ls(pattern="meta_"))
```


```{r env_data, echo=FALSE, warning = FALSE, fig.width=6,fig.height=4, fig.show="hold", out.width="50%"}
eval(parse("/Users/corahoerstmann/Documents/AWI_ArcticFjords/Code/env_calc/env_data.R", encoding = "UTF-8"))
list_meta$meta_all%>%group_by(Region)%>%summarize(n())
```



```{r alpha_div, echo=TRUE, fig.width=7,fig.height=3.5, fig.show="hold", out.width="50%"}
source("/Users/corahoerstmann/Documents/AWI_ArcticFjords/Code/alpha_div/alpha_diversity.R")

#the following is very intense computing so better to not run it too often
euk_r_alpha <- diversity_iNEXT(euk_r)
prok_alpha <- diversity_iNEXT(prokaryotes)


euk_r_alpha_meta <- left_join(euk_r_alpha, list_meta$meta_18S_r, by = "Site")
prok_alpha_meta <- left_join(prok_alpha, list_meta$meta_16S, by = "Site")

euk_r_alpha_n <- euk_r_alpha_meta%>%
  dplyr::rename(Richness.euk = Richness)%>%
  dplyr::rename(Shannon.euk = Shannon)%>%
  dplyr::rename(Simpson.euk = Simpson)

prok_alpha_n <- prok_alpha_meta%>%
  dplyr::rename(Richness.prok = Richness)%>%
  dplyr::rename(Shannon.prok = Shannon)%>%
  dplyr::rename(Simpson.prok = Simpson)

euk_r_alpha_n$Pielou.euk <- euk_r_alpha_n$Shannon.euk/log(euk_r_alpha_n$Richness.euk)
prok_alpha_n$Pielou.prok <- prok_alpha_n$Shannon.prok/log(prok_alpha_n$Richness.prok)


euk_r_alpha_meta_l <- euk_r_alpha_n%>%gather(Div_indices, Div_value, Richness.euk, Shannon.euk, Simpson.euk, Pielou.euk)
prok_alpha_meta_l <- prok_alpha_n%>%gather(Div_indices, Div_value, Richness.prok, Shannon.prok, Simpson.prok, Pielou.prok)
div_sub <- c("Station", "Glacial.influence","In.Out","Bioclimatic_subzone", "Div_indices", "Div_value")


#REF 4 pielou: https://www.davidzeleny.net/anadat-r/doku.php/en:div-ind



div_alpha_both <- full_join(euk_r_alpha_meta_l[,div_sub], prok_alpha_meta_l[,div_sub])

list_div <- sapply(ls(pattern="_alpha"), function(x) get(x), simplify = FALSE)


std <- function(x) sd(x)/sqrt(length(x))

list_div$div_alpha_both%>%filter(Div_indices == "Richness.euk")%>%group_by(Bioclimatic_subzone)%>%summarise(median = median(Div_value),min = min(Div_value), max = max(Div_value), std = std(Div_value), n=n())
list_div$div_alpha_both%>%filter(Div_indices == "Richness.prok")%>%group_by(Bioclimatic_subzone)%>%summarise(median = median(Div_value),min = min(Div_value), max = max(Div_value), std = std(Div_value), n=n())
rm(list = ls(pattern="_alpha"))

```



```{r z_scoring}
source("/Users/corahoerstmann/Documents/AWI_ArcticFjords/Code/data_transformations/z_scoring.R")
```

alpha diversity boxplots
```{r div_boxplots, fig.width=6,fig.height=4, out.width="50%"}
source("/Users/corahoerstmann/Documents/AWI_ArcticFjords/Code/alpha_div/boxplot_alphadiv.R")
```

## data transformations
```{r transformations, echo=TRUE, warning=FALSE}
source("/Users/corahoerstmann/Documents/AWI_ArcticFjords/Code/data_transformations/transformation_CLR_aitchinson.R")

euk_aitchinson <- dczm(euk_r, 1)
prok_aitchinson <- dczm(prokaryotes, 1)
#prok_aitchinson_r <- dczm(prok_r, 5)


euk_clr <- clr(euk_r,1)
prok_clr <- clr(prokaryotes, 1)

taxonomy_16S_clr <- taxonomy_16S%>%filter(rownames %in% rownames(prok_clr))
taxonomy_16S_clr$ASV <- taxonomy_16S_clr$rownames

```

```{r ait_dataframe, echo=FALSE, warning=FALSE}
euk_ait_df <- as.data.frame(as.matrix(euk_aitchinson))
prok_ait_df <- as.data.frame(as.matrix(prok_aitchinson))
#prok_ait_r_df <- as.data.frame(as.matrix(prok_aitchinson_r))
```

## multivariate analysis

**RDA**
```{r RDA, echo=FALSE, warning=FALSE, fig.width=6,fig.height=4, out.width="50%"}
source("/Users/corahoerstmann/Documents/AWI_ArcticFjords/Code/analysis_multivariate/RDA_plots.R")

source("/Users/corahoerstmann/Documents/AWI_ArcticFjords/Code/analysis_multivariate/RDA_ordistep.R")


## prokaryotes

RDA_input_16S <- c("Site","temperature...C.", "salinity..psu.", "Fluorometer",
                   "PO4_umol.l", "Si_umol.l", "NO3_umol.l", "MLD", "bottom_depth", "altitude", "azimuth")

prok_nNA <- prokaryotes%>%dplyr::select(meta_16S_nNA.z$Site)
prok_clr_nNA <- prok_clr%>%dplyr::select(meta_16S_nNA.z$Site)
prok_ait_nNA <- prok_ait_df%>%dplyr::select(meta_16S_nNA.z$Site)
prok_ait_nNA <- prok_ait_nNA%>%dplyr::filter(rownames(prok_ait_nNA) %in% colnames(prok_ait_nNA))


ait_dist_16S <- dczm(prok_nNA, 1)
ordi(prok_clr_nNA, meta_16S_nNA.z[RDA_input_16S])

#ordistep function removes Si

#RDA plot

meta_sig <- c("temperature...C.", "salinity..psu.",
                   "PO4_umol.l", "NO3_umol.l", "bottom_depth", "altitude")

RDA_plots(meta_16S_nNA.z, prok_clr_nNA, prok_ait_nNA)

#Plot against temperature gradient to show the beta diversity change

#RDA_plots_v2(meta_16S_nNA.z, prok_clr_nNA, prok_ait_nNA)

## eukaryotes

RDA_input_18S <- c("Site","Temperature...C.", "Salinity..psu.", "Fluorometer",
                   "PO4_umol.l", "Si_umol.l", "NO3_umol.l", "MLD", "bottom_depth", "altitude", "azimuth")



euk_clr_nNA <- euk_clr%>%dplyr::select(meta_18S_nNA.z$Site)
euk_ait_nNA <- euk_ait_df%>%dplyr::select(meta_18S_nNA.z$Site)
euk_nNA <- eukaryotes%>%dplyr::select(meta_18S_nNA.z$Site)

ait_dist_18S <- dczm(euk_nNA, 1)

ordi(euk_clr_nNA, meta_18S_nNA.z[RDA_input_18S])

#ordistep function confirms that all variables contribute significantly

meta_sig <- c("Temperature...C.", "Salinity..psu.", "Fluorometer",
                   "PO4_umol.l", "Si_umol.l", "NO3_umol.l", "bottom_depth", "altitude")
meta_18S_nNA.z$temperature...C. <- meta_18S_nNA.z$Temperature...C.
#RDA plot

RDA_plots(meta_18S_nNA.z, euk_clr_nNA, euk_ait_nNA)
#RDA_plots_v2(meta_18S_nNA.z, euk_clr_nNA, euk_ait_nNA)
```


## distance analysis

```{r distance, echo=FALSE, warning=FALSE, fig.width=6,fig.height=4, out.width="33%"}
source("/Users/corahoerstmann/Documents/AWI_ArcticFjords/Code/analysis_distance/distance_analyses.R")

distance_calc_prok(prok_r, info.sort, list_drifter$drifter_1m_prok, list_meta$meta_16S)
distance_calc_prok(prok_r, info.sort, list_drifter$drifter_3m_prok, list_meta$meta_16S)
distance_calc_prok(prok_r, info.sort, list_drifter$drifter_6m_prok, list_meta$meta_16S)
distance_calc_prok(prok_r, info.sort, list_drifter$drifter_1y_prok, list_meta$meta_16S)
distance_calc_prok(prok_r, info.sort, list_drifter$drifter_5y_prok, list_meta$meta_16S)

distance_calc_euk(euk_r, info.sort, list_drifter$drifter_1m_euk, list_meta$meta_18S)
distance_calc_euk(euk_r, info.sort, list_drifter$drifter_3m_euk,list_meta$meta_18S)
distance_calc_euk(euk_r, info.sort, list_drifter$drifter_6m_euk,list_meta$meta_18S)
distance_calc_euk(euk_r, info.sort, list_drifter$drifter_1y_euk,list_meta$meta_18S)
distance_calc_euk(euk_r, info.sort, list_drifter$drifter_5y_euk,list_meta$meta_18S)


##diversity against temperature gradients

source("/Users/corahoerstmann/Documents/AWI_ArcticFjords/Code/analysis_distance/Temperature_distance.R")

```

## functional co-analysis of prokaryotes and picoeukaryotes and their trophic groups

```{r network_analysis_1, echo=FALSE, warning=FALSE}
#concatenate the ASV tables after clr transformation

meta_all <- dplyr::inner_join(list_meta$meta_16S_m, list_meta$meta_18S_m[1:2], by="Station")

Prok_r_clr <- prok_clr%>%dplyr::select(meta_all$Site.x)
#rename the columns

Prok_r_clr_rename <- Prok_r_clr%>%rename_at(vars(meta_all$Site.x), ~ meta_all$Station)

Euk_r_clr <- euk_clr%>%dplyr::select(meta_all$Site.y)

Euk_r_clr_rename <- Euk_r_clr%>%rename_at(vars(meta_all$Site.y), ~ meta_all$Station)

#concatenate based on the same column names

merged_clr_all <- rbind(Prok_r_clr_rename, Euk_r_clr_rename)

#taxonomy

t_16S <- c("kingdom", "phylum", "class", "genus", "family", "species")
t_18S <- c("kingdom", "supergroup", "phylum", "class", "order", "family")
trans <- c("Tax_1", "Tax_2", "Tax_3", "Tax_4", "Tax_5", "Tax_6")

df <- data.frame(trans, t_16S, t_18S)

taxonomy_16S_r <- taxonomy_16S%>%dplyr::select(df$t_16S)
taxonomy_16S_r_rename <- taxonomy_16S_r%>%rename_at(vars(df$t_16S), ~ df$trans)

taxonomy_18S_r <- taxonomy_18S%>%dplyr::select(df$t_18S)
taxonomy_18S_r_rename <- taxonomy_18S_r%>%rename_at(vars(df$t_18S), ~ df$trans)

taxonomy_merged <- rbind(taxonomy_16S_r_rename, taxonomy_18S_r_rename)
taxonomy_merged_clr <- taxonomy_merged%>%dplyr::filter(rownames(taxonomy_merged) %in% rownames(merged_clr_all))

#replace "uncultured" with empty field

taxonomy_merged_clr$Tax_6 <- gsub("uncultured","",taxonomy_merged_clr$Tax_6)
taxonomy_merged_clr$Tax_5 <- gsub("uncultured","",taxonomy_merged_clr$Tax_5)
taxonomy_merged_clr$Tax_4 <- gsub("uncultured","",taxonomy_merged_clr$Tax_4)


taxonomy_merged_clr[taxonomy_merged_clr==" "]<-NA
taxonomy_merged_clr[taxonomy_merged_clr==""]<-NA

taxonomy_merged_clr$Tax_6[is.na(taxonomy_merged_clr$Tax_6)] <- "f__"
taxonomy_merged_clr$Tax_5[is.na(taxonomy_merged_clr$Tax_5)] <- "g__"
```



```{r trophic groups}
source("/Users/corahoerstmann/Documents/AWI_ArcticFjords/Code/analysis_network/top_abundant.R")
```
```{r BactoTrait, echo=FALSE, warning=FALSE}
source("/Users/corahoerstmann/Documents/AWI_ArcticFjords/Code/analysis_network/BactoTrait_annotation.R")
```

```{r PICRUST2 comparison, warning=FALSE}
source("/Users/corahoerstmann/Documents/AWI_ArcticFjords/Code/analysis_network/Picrust2_trait_comparison.R")
```
